status/controls.py

```diff
@@ -0,0 +1,110 @@
+        self.write( 
+            t.generate(
+                gs_globals=self.application.gs_globals, user=self.get_current_user(),
+                all_control_data = all_control_data, # control_data is a dictionary with the control data, it can be called in the html with the name before the equal sign
+                headers=headers,
+                ws_data = ws_data,
+                lims_uri=BASEURI,
+            )
+        )
+
+    def find_control_data(self, control_type):
+        """Find control data from the couchDB from project/controls view
+        """
+        from collections import defaultdict
+        result = defaultdict(dict)
+        controls_view = self.application.projects_db.view("project/controls")
+
+        all_cont_proj = controls_view[[control_type,'']:[control_type,"Z"]]
+        for cont_proj in all_cont_proj:
+            for cont_sample in cont_proj.value:
+                for workset in cont_proj.value[cont_sample]: # here we create one entry in result for each workset, this will be one line in the controls table
+                    if workset != "no_workset":
+                        result[workset]["sample_id"] = cont_sample
+                        result[workset]["customer_name"] = cont_proj.value[cont_sample][workset]["customer_name"]
+                        if "status_manual" in cont_proj.value[cont_sample][workset]: # status originates from LIMS project overview, is often not set for controls
+                            result[workset]["status_manual"] = cont_proj.value[cont_sample][workset]["status_manual"]
+                        else:
+                            result[workset]["status_manual"] = "* In Progress" # asterisk indicates that the status in LIMS is not set, the sample has a workset and so MUST be at least "In Progress"
+                        result[workset]["project"] = cont_proj.key[1]
+                        result[workset]["workset_name"] = cont_proj.value[cont_sample][workset]["workset_name"]
+                        if "workset_id" in cont_proj.value[cont_sample][workset]:
+                            result[workset]["workset_id"] = cont_proj.value[cont_sample][workset]["workset_id"]
+                        else:
+                            result[workset]["workset_id"] = "NA"
+                        if "prep_status" in cont_proj.value[cont_sample][workset]:
+                            result[workset]["prep_status"] = cont_proj.value[cont_sample][workset]["prep_status"]
+                        else:                            
+                            result[workset]["prep_status"] = "" 
+                        result[workset]["sequenced_fc"] = cont_proj.value[cont_sample][workset]["sequenced_fc"]
+        return result
+    
+    def worksets_data(self):
+        """retrieves projects for each workset and return a dictionary:
+        {[workset_id, workset_name]: {'ws_project_names':[P.roject_00_01, P.roject_00_02, ...], 'ws_project_numbers':[P12346, P23456, ...]}}
```

Comment:
Consider revising since the data structure looks different now?

Reply:
ahhhh, right. You mean the comment, right?